### PR TITLE
feat: ecocredit ORM proto definitions 

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

adds ORM proto definitions for x/ecocredit

Closes: n/a

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    // metadata is any arbitrary metadata attached to the project.
+    bytes metadata = 5;
+}
+
+// BatchInfo represents the high-level on-chain information for a credit batch.
+message BatchInfo {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 3,
+        primary_key: {fields: "batch_denom"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:67]: Let's actually add an uint64 auto increment primary key and make this an unique index. I think we'll want a short uint64 to reference batches in the order book and denoms are getting kind of long

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import "cosmos/base/v1beta1/coin.proto";
+import "gogoproto/gogo.proto";
+import "google/protobuf/timestamp.proto";
+
+// ClassInfo represents the high-level on-chain information for a credit class.
+message ClassInfo {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "class_id"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:15]: Maybe let's add admin and credit type indexes?

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    // The number of batches issued in this credit class.
+    uint64 num_batches = 6;
+}
+
+// ProjectInfo represents the high-level on-chain information for a project.
+message ProjectInfo {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "project_id"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:42]: How about issuer and class indexes?

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    // the measurement unit (e.g. kg, ton, etc)
+    string unit = 3;
+
+    // the decimal precision
+    uint32 precision = 4;
+}
+
+// SellOrder represents the information for a sell order.
+message SellOrder {
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:126]: Let's move orders into a separate proto package

**Review by aaronc — COMMENTED** [review]
Great start @technicallyty ! Left a few suggestions

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    };
+
+    // class_id is the unique ID of credit class.
+    string class_id = 1;
+
+    // admin is the admin of the credit class.
+    string admin = 2;
+
+    // issuers are the approved issuers of the credit class.
+    repeated string issuers = 3;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:25]: In keeping with 1NF (which I think we should), let's move this to its own join table

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 1 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:67]: are there docs anywhere for the syntax on this?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 4 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto]: i separated it like:

`ecocredit/v1beta1/marketplace/state.proto`


is that how we want it?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 2 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:15]: https://github.com/regen-network/regen-ledger/pull/700/files#diff-302ae7137e2a8ea4a04301cf72f3df8311c99ea4716efaf59ae04291e030fb63R16-R17

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 3 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:55]: added: https://github.com/regen-network/regen-ledger/pull/700/files#diff-302ae7137e2a8ea4a04301cf72f3df8311c99ea4716efaf59ae04291e030fb63R56-R57

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import "cosmos/orm/v1alpha1/orm.proto";
+import "cosmos/base/v1beta1/coin.proto";
+
+// SellOrder represents the information for a sell order.
+message SellOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "order_id"}
+        index: {id: 1 fields: "batch_denom"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/marketplace/state.proto:14]: how about owner index?

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    // resell the credits to another buyer.
+    bool disable_auto_retire = 6;
+}
+
+// BuyOrder represents the information for a buy order.
+message BuyOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "buy_order_id"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/marketplace/state.proto:45]: buyer index too?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 1 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto:82]: in orm.proto
```suggestion
        primary_key: {fields: "id", auto_increment: true}
        index: {fields: "batch_denom", unique: true}
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    // The number of batches issued in this credit class.
+    uint64 num_batches = 5;
+}
+
+// ClassIssuers is a JOIN table for Class Info that stores the credit class issuers
+message ClassIssuers {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "class_id"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:40]:
```suggestion
        primary_key: {fields: "class_id,issuer"}
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    };
+
+    // class_id is the unique ID of credit class.
+    string class_id = 1;
+
+    // issuers are the approved issuers of the credit class.
+    repeated string issuers = 2;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:47]:
```suggestion
   string issuer = 2;
```

in 1NF we can't have any `repeated` fields

**Review by aaronc — COMMENTED** [review]
Looking good. Don't we have sequences to track for generating class, project and batch IDs?

**Comment by technicallyty** [comment]
> Looking good. Don't we have sequences to track for generating class, project and batch IDs?

right, so these would be separate messages/tables right? i think batches are covered now with the addition of the auto_increment id

**Comment by aaronc** [comment]
> 

Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 4 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto]: Yep 👍

**Comment by aaronc** [comment]
We also need balances and supply

**Comment by technicallyty** [comment]
> Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?

actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?

i've added a project seq table, and credit/basket balance. looks like the supply is already tracked in `BatchInfo` via `total_amount`.

**Comment by aaronc** [comment]
> > Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?
> 
> actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?
> 
> i've added a project seq table, and credit/basket balance. looks like the supply was already tracked in `BatchInfo` via `total_amount`.

I sort of have the preference that things that get updated somewhat frequently - like sequences and supply - should live on separate tables from things which rarely or never change. It's not a huge deal, but I want to avoid the pattern that is done some places in the SDK like x/auth where every acct seq update rewrites the whole account with vesting, pub key and all

**Comment by technicallyty** [comment]
> > > Yes separate tables, but I think not quite for batch ID because these batch IDs are unique per project and project IDs unique per class right?
> > 
> > 
> > actually looks like we were tracking batch sequence in the class via the field `num_batches`. should we continue with this method?
> > i've added a project seq table, and credit/basket balance. looks like the supply was already tracked in `BatchInfo` via `total_amount`.
> 
> I sort of have the preference that things that get updated somewhat frequently - like sequences and supply - should live on separate tables from things which rarely or never change. It's not a huge deal, but I want to avoid the pattern that is done some places in the SDK like x/auth where every acct seq update rewrites the whole account with vesting, pub key and all

ok, added batch sequence and separate batch supply table

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+    // the measurement unit (e.g. kg, ton, etc)
+    string unit = 3;
+
+    // the decimal precision
+    uint32 precision = 4;
+}
+
+// CreditTypeSequence stores and increments the sequence number for credit types
+message CreditTypeSequence {
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:129]: What are these sequences being used for?

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+// ProjectSequence stores and increments the sequence number for projects
+message ProjectSequence {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 7,
+        primary_key: {fields: "project_id"}
+    };
+
+    string project_id = 1;
+
+    uint64 id = 2;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:149]:
```suggestion
    uint64 next_id = 2;
```

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:129]: i believe it was added in the case we want multiple types of "carbon" credits with different parameters, i.e carbon but measured in a different unit. in which case the abbreviation would still contain "C" and then a sequence for each additional type C0, C01, C02 etc

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    // batch_denom is the denom of the credit batch
+    string batch_denom = 2;
+
+    message CreditBalance {
+        string tradable = 1;
+        string retired = 2;
+    }
+
+    // balance is the balance of both tradable and retired credits
+    CreditBalance balance = 3;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:171]:
```suggestion
        string tradable = 1;
        string retired = 2;
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
+
+    // balance is the amount of ecocredits held in the basket
+    string balance = 3;
+}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:189]: Let's put this in a basket sub-package

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto:129]: that doesn't make sense 🤔 . I think we should nix it. the abbreviation is the primary key and this is controlled by governance

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
+        primary_key: {fields: "basket_denom,batch_denom"}
+    };
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:180]: can we reduce storage space using uint64 id's?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:129]: would it still be auto-generated then?

Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto:129]: credit type was never meant to be auto-generated. it is chosen by governance. i'm confused by where this seq came from tbh

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    };
+
+    // id is the id of the batch
+    uint64 id = 1;
+
+    // total_amount is the total number of active credits in the credit batch.
+    // Some of the issued credits may be cancelled and will be removed from
+    // total_amount and tracked in amount_cancelled. total_amount and
+    // amount_cancelled will always sum to the original credit issuance amount.
+    string total_amount = 2;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:202]: Where are we tracking retired amount? That's different from cancelled

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto:129]: no, the abbreviation is chosen by governance. just `C` for instance. let's fix it

File: proto/regen/ecocredit/v1beta1/baskets/state.proto [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
+        primary_key: {fields: "basket_denom,batch_denom"}
+    };
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the batch
+    string batch_denom = 2;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/baskets/state.proto:18]:
```suggestion
    uint64 batch_id = 2;
```

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 15 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:202]: it wasn't tracked in the original design. i can split the supply into tradable and retired

Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto]: https://github.com/regen-network/regen-ledger/pull/700/commits/40bbe1ed417b578073d2f232230562cd54a61cf5

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto]: https://github.com/regen-network/regen-ledger/pull/700/commits/a40947c8ec4f81850b3491bb44aab356be08f474

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 4,
+        primary_key: {fields: "id", auto_increment: true}
+        index: {fields: "batch_denom", unique: true}
+        index: {id: 1 fields: "project_id"}
+        index: {id: 2 fields: "start_date"}
+    };
+
+    // id is a an incrementing integer to identify the batch
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:85]:
```suggestion
    // id is an auto-incrementing integer to succintly identify the batch
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+// EcocreditBalance stores each users credit balance
+message EcocreditBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 8,
+        primary_key: {fields: "address,batch_id"}
+    };
+
+    // address is the address of the credit holder
+    string address = 1;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:150]: let's use `bytes`

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 10 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto]: added an issue to remove the sequencing logic from the ecocredit module https://github.com/regen-network/regen-ledger/issues/707

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+    uint64 batch_id = 2;
+
+    // tradable is the tradable amount of credits
+    string tradable = 3;
+
+    // retired is the retired amount of credits
+    string retired = 4;
+}
+
+// BatchSequence tracks the sequence number for batches within a class
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:162]:
```suggestion
// BatchSequence tracks the sequence number for batches within a project
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+message BatchSequence {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 10,
+        primary_key: {fields: "class_id"}
+    };
+
+    // class_id is the id of the class for a batch
+    uint64 class_id = 1;
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:170]:
```suggestion
    string project_id = 1;
```

**Review by aaronc — COMMENTED** [review]
getting close! thanks for your patience!

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 18 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto]: should we be using cosmos scalars? or just change to bytes for now

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+
+    // next_id is the sequence number for the project
+    uint64 next_id = 2;
+}
+
+// EcocreditBalance stores each users credit balance
+message EcocreditBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 8,
+        primary_key: {fields: "address,batch_id"}
```
Comment (aaronc) [inline proto/regen/ecocredit/v1beta1/state.proto:146]:
```suggestion
        primary_key: {fields: "address,batch_id"}
        index: {id: 1, fields: "batch_id,address"}
```

**Review by aaronc — APPROVED** [review]
LGTM. Thanks @technicallyty 🎉 !

**Comment by aaronc** [comment]
Just need to fix conflicts

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+// SellOrder represents the information for a sell order.
+message SellOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 1,
+        primary_key: {fields: "order_id"}
+        index: {id: 1 fields: "batch_denom"}
+        index: {id: 2 fields: "owner"}
+    };
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1beta1/marketplace/state.proto:16]: Following #671, we need to include the `expiration` message field and add an expiration index.

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// BuyOrder represents the information for a buy order.
+message BuyOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "buy_order_id"}
+        index: {id: 1 fields: "buyer"}
+    };
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1beta1/marketplace/state.proto:48]: Following #671, we need to include the expiration message field and add an expiration index.

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    // the credits will not auto-retire and can be resold assuming that the
+    // corresponding sell order has auto-retirement disabled. If the sell order
+    // hasn't disabled auto-retirement and the buy order tries to disable it,
+    // that buy order will fail.
+    bool disable_auto_retire = 6;
+
+    // disable_partial_fill disables the default behavior of partially filling
+    // buy orders if the requested quantity is not available.
+    bool disable_partial_fill = 7;
+}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1beta1/marketplace/state.proto:98]: We need to add `AllowAskDenom` here as well.

File: proto/regen/ecocredit/v1beta1/baskets/state.proto [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
+        primary_key: {fields: "basket_denom,batch_id"}
+    };
+
+    // basket_denom is the denom of the basket
+    string basket_denom = 1;
+
+    // batch_id is the denom of the batch
+    uint64 batch_id = 2;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1beta1/baskets/state.proto:18]:
```suggestion
    // batch_id is the id of the credit batch
    uint64 batch_id = 2;
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 26]
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
...
+    string tradable = 3;
+
+    // retired is the retired amount of credits
+    string retired = 4;
+}
+
+// BatchSequence tracks the sequence number for batches within a project
+message BatchSequence {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 10,
```
Comment (clevinson) [inline proto/regen/ecocredit/v1beta1/state.proto:166]:
```suggestion
        id: 9,
```

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 27]
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
...
+    string project_id = 1;
+
+    // num_batches is a sequence number incrementing on each issued batch
+    uint64 num_batches = 2;
+}
+
+// BatchSupply tracks the supply of a credit batch
+message BatchSupply {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 11,
```
Comment (clevinson) [inline proto/regen/ecocredit/v1beta1/state.proto:180]:
```suggestion
        id: 10,
```

File: proto/regen/ecocredit/v1beta1/baskets/state.proto [hunk 28]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package regen.ecocredit.v1beta1.baskets;
+
+import "cosmos/orm/v1alpha1/orm.proto";
+
+// BasketBalance stores the amount of credits from a batch in a basket
+message BasketBalance {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 9,
```
Comment (clevinson) [inline proto/regen/ecocredit/v1beta1/baskets/state.proto:10]:
```suggestion
        id: 1,
```

Comment (clevinson) [reply proto/regen/ecocredit/v1beta1/baskets/state.proto:10]: Should this be `1`? As its the only table in this file

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 24 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:127]: https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 23 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:53]: https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 22 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:17]: https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9

Comment (ryanchristo) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:17]: Looks like the expiration index is still missing..

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 23 above)
Comment (ryanchristo) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:53]: Same here..

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 22 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto]: oh whoops, think i missed the index part and only added it as a field 😅 its in there now!

**Review by ryanchristo — APPROVED** [review]

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+
+    // Selection defines a buy order selection.
+    message Selection {
+
+        // sum defines the type of selection.
+        oneof sum {
+            // sell_order_id is the sell order ID against which the buyer is trying to buy.
+            // When sell_order_id is set, this is known as a direct buy order because it
+            // is placed directly against a specific sell order.
+            uint64 sell_order_id = 1;
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:65]: i think that would be out of the scope of this PR. this structure was copied from existing proto defs. so i believe this concern would warrant its own issue

File: proto/regen/ecocredit/v1beta1/state.proto [hunk 30]
```diff
@@ -3,25 +3,198 @@ syntax = "proto3";
...
+        id: 7,
+        primary_key: {fields: "project_id"}
+    };
+
+    // project_id is the id of the project
+    string project_id = 1;
+
+    // next_id is the sequence number for the project
+    uint64 next_id = 2;
+}
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply proto/regen/ecocredit/v1beta1/state.proto:140]: so the format of the project id is classID + sequence number. i think actually the project_id field should be class_id. but in any case, the 2 fields are needed

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 29 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:65]: why a list?

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 30 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/state.proto:140]: it should be class_id

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 29 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:65]: @aaronc to select multiple orders to fill.

File: proto/regen/ecocredit/v1beta1/state.proto (hunk 30 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/regen/ecocredit/v1beta1/state.proto:140]: ah, so `ProjectSequence` is a type for a key.

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/regen/ecocredit/v1beta1/state.proto:140]: @technicallyty if you will be updating the `project_id` field, let's update the `message` documentation as well.

File: proto/regen/ecocredit/v1beta1/marketplace/state.proto (hunk 29 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply proto/regen/ecocredit/v1beta1/marketplace/state.proto:65]: this model is just based on buying from a specific sell order. it's not a batch operation. it's like a shopping cart where you buy one product from one vendor

**Pull request merged**